x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/transport/netty4/SecurityNetty4HttpServerTransportCloseNotifyTests.java

```diff
@@ -0,0 +1,284 @@
+            try (var ctx = connectClientAndServer(tlsVersion)) {
+                ctx.client.channel.writeAndFlush(new DefaultFullHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.GET, "/index"));
+                var serverRequestCtx = safePoll(ctx.server.dispatcher.reqQueue);
+                serverRequestCtx.restChannel.sendResponse(new RestResponse(RestStatus.OK, ""));
+                safePoll(ctx.client.respQueue);
+                var ssl = ctx.client.channel.pipeline().get(SslHandler.class);
+                ssl.closeOutbound();
+                safeAwait(ctx.client.channel.closeFuture());
+            }
+        });
+    }
+
+    /**
+     * This test ensures that sending close_notify with outstanding requests close channel immediately.
+     */
+    public void testSendCloseNotifyBeforeHttpResponse() {
+        runForAllTlsVersions(tlsVersion -> {
+            try (var ctx = connectClientAndServer(tlsVersion)) {
+                var server = ctx.server;
+                var client = ctx.client;
+
+                var nRequests = randomIntBetween(1, 5);
+                for (int i = 0; i < nRequests; i++) {
+                    client.channel.writeAndFlush(new DefaultFullHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.GET, "/index"));
+                }
+                assertBusy(() -> assertEquals(nRequests, server.dispatcher.reqQueue.size()));
+
+                // after the server receives requests send close_notify, before server responses
+                var ssl = client.channel.pipeline().get(SslHandler.class);
+                ssl.closeOutbound();
+
+                safeAwait(ctx.client.channel.closeFuture());
+                assertTrue(client.errQueue.isEmpty());
+                assertTrue(client.respQueue.isEmpty());
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+        });
+
+    }
+
+    private record HttpServer(Netty4HttpServerTransport netty, QueuedDispatcher dispatcher, ThreadPool threadPool) {}
+
+    private record HttpClient(
+        Bootstrap netty,
+        Channel channel,
+        BlockingDeque<FullHttpResponse> respQueue,
+        BlockingDeque<Throwable> errQueue
+    ) {}
+
+    private record ConnectionCtx(String tlsProtocol, HttpServer server, HttpClient client) implements AutoCloseable {
+
+        @Override
+        public void close() {
+            server.netty.stop();
+            server.threadPool.shutdownNow();
+            safeAwait(client.netty.config().group().shutdownGracefully());
+        }
+    }
+
+    private static class QueuedDispatcher implements HttpServerTransport.Dispatcher {
+        BlockingQueue<ReqCtx> reqQueue = new LinkedBlockingDeque<>();
+        BlockingDeque<ErrCtx> errQueue = new LinkedBlockingDeque<>();
+
+        @Override
+        public void dispatchRequest(RestRequest request, RestChannel channel, ThreadContext threadContext) {
+            reqQueue.add(new ReqCtx(request, channel, threadContext));
```

Comment:
I have not investigated the test failure much. But I'm pretty sure these are going to need to be released at the end of the test.

Reply:
Or else the leak detector will fire.